### PR TITLE
fix: harden webhook/link/OAuth-avatar SSRF (advisory clusters A/B/C/E)

### PR DESCRIPTION
**Pull request by sriramveeraghanta**
### Description

Resolves three overlapping SSRF advisory clusters affecting **webhook delivery** and **work-item link unfurling**. The root fix is a single shared pinned-IP HTTP client so the address we validate is exactly the address we connect to.

- **Cluster A — private-IP validation + PATCH bypass**
  - `WebhookEndpoint.patch` passed `context={request: request}` — the `request` *object* as the dict key — so `self.context.get("request")` was always `None` and the loopback/disallowed-domain guard silently no-op'd on update. Fixed to `context={"request": request}` (GHSA-6485-m23r-fx8q).
  - Hardened IP classification (`is_blocked_ip`) to also block **multicast**, **unspecified**, **CGNAT `100.64.0.0/10`** (not `is_private` on Python 3.12), and IPv4 addresses **embedded in IPv6 transition formats** (IPv4-mapped `::ffff:`, NAT64, 6to4, Teredo). Fails closed and is robust across Python versions rather than relying on 3.12-specific `ipaddress` semantics.

- **Cluster B — DNS-rebinding TOCTOU (the previously-unpatched core)**
  - Validators resolved DNS, then `requests` resolved it **again** at connect time, letting an attacker swap in an internal IP between the two lookups.
  - New `plane/utils/url_security.py` resolves + validates **once**, then connects to the **validated IP literal** so urllib3 performs no second DNS lookup (TOCTOU closed by construction), while preserving the original hostname for the `Host` header, TLS SNI, and certificate verification (`PinnedIPAdapter` injects `server_hostname`; `assert_hostname` left at `None` so cert checking stays on).

- **Cluster C — SSRF via HTTP redirect following**
  - Webhook delivery never auto-follows redirects (`allow_redirects=False`); the link crawler follows them manually, re-resolving + re-validating + **re-pinning every hop** (a relative `Location` is resolved against the logical hostname URL, not the IP-literal URL).

Implementation notes:
- Pinned `requests==2.33.0` in `base.txt` (it is imported directly by production code, and the pinning adapter needs the `>=2.32` `get_connection_with_tls_context` hook).
- `trust_env=False` + proxies disabled on the pinned session so an ambient `HTTP(S)_PROXY` can't tunnel around the pin.
- Trusted `WEBHOOK_ALLOWED_HOSTS` entries skip the *block* check but are **still pinned**, so a trusted hostname can't be rebound to a different internal target.
- Webhook URL-validation rejections are now recorded in `WebhookLog` (visible to admins) instead of being silently swallowed.

This work was hardened by an adversarial review pass (independent attacker/verifier agents) which caught and fixed: the `allowed_hosts` path missing pinning, a NAT64 prefix-mask inconsistency, and silent `ValueError` handling.

### Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (change that would cause existing functionality to not work as expected)

### Screenshots and Media (if applicable)

N/A — backend SSRF hardening, no UI changes.

### Test Scenarios

- **Webhook to a public endpoint** still delivers, with `Host`/TLS/cert verified against the real hostname (connection pinned to the resolved IP).
- **Webhook whose URL resolves to an internal/blocked IP** (e.g. `127.0.0.1`, `169.254.169.254`, `100.64.0.1`, `::ffff:169.254.169.254`) is rejected and the rejection is recorded in `WebhookLog`.
- **DNS-rebinding**: a hostname returning a public IP at validation and a private IP at connect time still connects only to the validated IP.
- **Webhook PATCH** updating the URL re-applies the loopback/disallowed-domain guard (previously bypassed via the bad context key).
- **Work-item link unfurl** to a public URL fetches title/favicon; a link (or redirect hop) pointing to a private IP is blocked; a redirect loop hits the hop cap.
- Unit suite: `docker compose -f docker-compose-test.yml run --rm api-tests pytest -m unit` → **178 passed**. New `test_url_security.py` covers pinning, rebinding, redirect re-validation, IP edge cases, and TLS SNI injection.

### References

- Advisory clusters A / B / C (webhook & link SSRF), related to CVE-2026-30242.
- GHSA-6485-m23r-fx8q (PATCH serializer context-key bypass).
- Cluster B GHSAs: GHSA-whh3-5g95-4qhc, GHSA-4mjx-q738-87cf, GHSA-6p39-x6q9-h3g5, GHSA-9292-pvg4-7hvm, GHSA-fv24-3845-646g, GHSA-8wvv-p676-hcw4.
- Cluster C GHSAs: GHSA-6v37-328w-j2wv, GHSA-jw6g-h7h5-rfc6, GHSA-mq87-52pf-hm3h.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed serializer context handling for PATCH webhook requests.

* **Security Improvements**
  * Hardened SSRF protections for outbound HTTP: DNS-pinning, per-hop validation, and safer redirect handling for webhooks, link previews, favicon probes, and OAuth avatar downloads.
  * Webhook delivery now rejects invalid webhook URLs early with clear logging and avoids unsafe network calls.

* **Chores**
  * Pinned HTTP client dependency for consistent network/TLS behavior.

* **Tests**
  * Added comprehensive unit tests covering URL/SSRF defenses and related flows.

[Review Change Stack](https://app.coderabbit.ai/change-stack/makeplane/plane/pull/9163?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by sriramveeraghanta** [comment]
## SSRF advisory verification (all similar cases checked)

I pulled every SSRF-class advisory from the repo's security advisories and verified each against the code on this branch. Summary: the webhook/link/favicon reports — including the ones users are re-filing as *"still bypassable"* — are resolved here, and the one genuinely-unresolved family (**OAuth avatar SSRF**) is now fixed too.

| Advisory | Class | Status on this branch |
|---|---|---|
| GHSA-m3f8-q4wj-9grv (closed) / **CVE-2026-30242** / GHSA-75vf-hh93-h7mx | webhook → internal IP | ✅ blocked (hardened `validate_url` + pinned send) |
| GHSA-75fg-f8qg-23wv | webhook CGNAT/6to4/multicast missed | ✅ now blocked (`is_blocked_ip`) |
| GHSA-6485-m23r-fx8q | webhook PATCH context-key bypass | ✅ fixed (`context={"request": request}`) |
| GHSA-whh3 / -4mjx / -6p39 / -9292 / **-fgcv-6h3f-xcx9** | webhook DNS-rebinding TOCTOU | ✅ closed (IP pinning) |
| GHSA-6v37-328w-j2wv / -jw6g-h7h5-rfc6 / -mq87-52pf-hm3h | webhook redirect SSRF | ✅ closed (`allow_redirects=False`) |
| GHSA-8wvv-p676-hcw4 / -fv24-3845-646g / -9292 | link-unfurl DNS rebinding | ✅ closed (pinned `safe_get`) |
| GHSA-9fr2-pprw-pp9j / **CVE-2026-39843** | favicon redirect SSRF | ✅ blocked (per-hop re-validation) |
| **GHSA-3856-6mgg-rx84** | favicon DNS rebinding (bypass of CVE-2026-39843) | ✅ closed (pinning) |
| **GHSA-cv9p-325g-wmv5** | OAuth avatar redirect SSRF → static-asset exfil | ✅ **fixed here** (avatar fetch now pinned) |
| **GHSA-hx79-5pj5-qh42** | Gitea OAuth SSRF | ⚠️ avatar hop fixed; `GITEA_HOST` first-hop intentionally **not** IP-blocked (see note) |

**Note on GHSA-hx79 first-hop:** `GITEA_HOST` is admin-set config, and a self-hosted Gitea on a private network is a legitimate deployment, so IP-blocking the token/userinfo requests would break valid setups. The exploitable, low-privilege vector in that advisory is the **avatar** (any Gitea user sets a malicious avatar URL), which is now blocked for all providers via `download_and_upload_avatar` → `pinned_fetch_following_redirects`.

### Tests
`apps/api/plane/tests/unit/bg_tasks/test_ssrf_advisories.py` adds a per-advisory regression map (each test reproduces the advisory scenario and asserts it is blocked). Full unit suite via `docker compose -f docker-compose-test.yml`: **199 passed**.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: apps/api/plane/utils/url_security.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        return self.poolmanager.connection_from_host(**host_params, pool_kwargs=pool_kwargs)
+
+
+def _split_target(url):
+    """Parse a URL into the pieces needed to build a pinned request."""
+    parts = urlsplit(url)
+    scheme = parts.scheme
+    if scheme not in ("http", "https"):
+        raise ValueError("Invalid URL scheme. Only HTTP and HTTPS are allowed")
+    hostname = parts.hostname
```
[the remark this answers is not in the document]
Comment (sriramveeraghanta) [reply apps/api/plane/utils/url_security.py:90]: Fixed in 7932bd4 — `_split_target` now preserves URL-embedded credentials and passes them through as `auth=(user, pass)` (percent-decoded), so Basic Auth still works after the URL is rewritten to the IP literal.

File: apps/api/plane/utils/url_security.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    """Issue a single request whose socket is pinned to ``ip``."""
+    ip_obj = ipaddress.ip_address(ip)
+    host_for_url = f"[{ip}]" if ip_obj.version == 6 else ip
+    url = f"{scheme}://{host_for_url}:{port}{path}"
+
+    request_headers = dict(headers or {})
+    default_port = 443 if scheme == "https" else 80
+    # Host header (and TLS) carry the ORIGINAL hostname, not the IP literal.
+    request_headers["Host"] = hostname if port == default_port else f"{hostname}:{port}"
+
```
[the remark this answers is not in the document]
Comment (sriramveeraghanta) [reply apps/api/plane/utils/url_security.py:120]: Fixed in 7932bd4 — IPv6-literal hostnames are now bracketed in the `Host` header (`Host: [2606:...]`).

File: apps/api/plane/utils/ip_address.py [hunk 3]
```diff
@@ -8,10 +8,160 @@
...
+
+    Raises:
+        ValueError: If the hostname cannot be resolved or (when
+        ``require_safe``) any resolved address is a blocked/internal target not
+        covered by ``allowed_ips``.
+    """
+    try:
+        addr_info = socket.getaddrinfo(hostname, None)
+    except socket.gaierror:
+        raise ValueError("Hostname could not be resolved")
```
[the remark this answers is not in the document]
Comment (sriramveeraghanta) [reply apps/api/plane/utils/ip_address.py:141]: Fixed in 7932bd4 — `resolve_and_validate` (and `validate_url_ip`) now catch `(socket.gaierror, UnicodeError)`, so IDNA failures surface as `ValueError` and get recorded as URL rejections.

File: apps/api/plane/authentication/adapter/base.py [hunk 4]
```diff
@@ -146,8 +146,16 @@ def download_and_upload_avatar(self, avatar_url, user):
...
+            # Download the avatar image over an SSRF-safe client: the avatar URL
+            # comes from the OAuth provider's (attacker-influenceable) profile
+            # data, so it must not be allowed to reach internal addresses. The
+            # connection is pinned to the validated IP (defeats DNS rebinding)
+            # and every redirect hop is re-validated, so a public URL cannot
+            # bounce the fetch to an internal target — GHSA-cv9p-325g-wmv5 /
+            # GHSA-hx79-5pj5-qh42 (avatar hop).
+            response, _ = pinned_fetch_following_redirects(
+                "GET", avatar_url, headers=headers, timeout=10, max_redirects=5
+            )
```
[the remark this answers is not in the document]
Comment (sriramveeraghanta) [reply apps/api/plane/authentication/adapter/base.py:161]: Fixed in 7932bd4 — the avatar download now uses `stream=True` (the pinned client keeps the session open until the response is closed), uploads the size-bounded `content` buffer instead of `response.content`, and always closes the response.

File: apps/api/plane/tests/unit/bg_tasks/test_ssrf_advisories.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+  * GHSA-8wvv-p676-hcw4 / -fv24-3845-646g / -9292-pvg4-7hvm  link rebinding
+  * GHSA-9fr2-pprw-pp9j / CVE-2026-39843  favicon redirect SSRF        -> TestFaviconRedirect
+  * GHSA-3856-6mgg-rx84  favicon DNS-rebinding                         -> TestFaviconRebinding
+
+OAuth avatar (the still-unresolved family this change adds)
+  * GHSA-cv9p-325g-wmv5  OAuth avatar redirect SSRF -> static-asset exfil
+  * GHSA-hx79-5pj5-qh42  Gitea OAuth SSRF (avatar hop)                 -> TestOAuthAvatarSSRF
+"""
+
+import ipaddress
```
[the remark this answers is not in the document]
Comment (sriramveeraghanta) [reply apps/api/plane/tests/unit/bg_tasks/test_ssrf_advisories.py]: Fixed in 7932bd4 — removed the unused `import ipaddress`.

**Review by pushya22 — APPROVED** [review]

**Review by Prashant-Surya — APPROVED** [review]
